### DISCUSSION TITLE
misaligned tokens on debug data

### DISCUSSION THREAD

**Discussion by mehmetilker** [explosion/spaCy#8564] [Help: Coding & Implementations]
I am creating training corpus with DocBin and exclude problematic annotations with the following way:

```
                    entity = doc.char_span(start, end, label=label, alignment_mode="strict")
                    if entity is not None:
                        ...
```

So I expect there won't be any problem with my training data but I see following warnings for 21000 training data:

⚠ 2658 misaligned tokens in the training data
✘ 1 invalid whitespace entity spans
⚠ 5 entity span(s) with punctuation

Especially for the first warning, reason could be tokenization as I customized tokenization due to token like this: ``` 'word ```

I have added same tokenization code to the debug data cli with following function:

```
@registry.callbacks("customize_tokenizer")
def make_customize_tokenizer():
    "https://spacy.io/usage/training#custom-tokenizer"
    def customize_tokenizer(nlp):
        prefixes = nlp.Defaults.prefixes + [r'''^[']''', ]
        prefixes_re = spacy.util.compile_prefix_regex(prefixes)

        def custom_tokenizer(nlp: Language):
            return Tokenizer(nlp.vocab,
                             rules=nlp.tokenizer.rules,
                             prefix_search=prefixes_re.search,
                             suffix_search=nlp.tokenizer.suffix_search,
                             infix_finditer=nlp.tokenizer.infix_finditer,
                             url_match=nlp.tokenizer.url_match)

        nlp.tokenizer = custom_tokenizer(nlp)

        return nlp
    return customize_tokenizer

```

But same warnings appear. What the reason could be that I do not see any problem on creating char_span but cannot pass "debug data" ?

**Comment by adrianeboyd** [comment]
For NER, the misalignments may not matter that much as long as the start of the entity and the end of the entity land on token boundaries in both the gold annotation and using the tokenizer in the config. It's more of an issue for tagging and parsing, where each individual token alignment matters.

If you use the same tokenizer for the `DocBin` and in your config, you shouldn't see misalignments, though, so maybe something isn't configured like you expect?

The only really concerning error is the whitespace entity span, which is definitely going to cause errors while training, so try to find that and filter it out.

You should also definitely have a look at how many entities aren't aligning (are `None` with `char_span`) to be sure that you're not losing a lot of your annotations before training/evaluating. Double-check the entity spans in the DocBin docs to make sure that the offsets are correct and the entity spans are actually the spans from the original training data, since it can be easy to accidentally end up with off-by-one errors for offsets in conversions.

**Reply by mehmetilker** [reply]
@adrianeboyd thanks for your advises.

I could solve error but no success to find out misaligned entities:

I tried to replicate "data debug" with following code.

```python
    from spacy.tokens import DocBin
    from spacy.training import Example
    nlp = customSpaCy.load_spacy_plain(configHelper.get_nlp_model_name())

    db = DocBin().from_disk(spacy_loc)
    for doc in db.get_docs(nlp.vocab):
        example = Example(nlp.make_doc(doc.text), doc)
        align = example.alignment
        data = {}
        data["n_misaligned_words"] = 0
        for token in doc:
            if token.orth_.isspace():
                continue
            if align.x2y.lengths[token.i] != 1:
                data["n_misaligned_words"] += 1
        if data["n_misaligned_words"] > 0:
            print(doc, data["n_misaligned_words"])
```

I expected that it will show me which docs contains misaligned entities.
I guess ```example = Example(nlp.make_doc(doc.text), doc)``` does not work as expected.

When I cancel tokenization on corpus preparation process, I see few errors. So it works as expected. In this case I should add same tokenization rule to the "debug data" command.

But problem  (I guess) is custom tokenization function does not work as expected. when I change that method I do not see any difference.
For example here "https://spacy.io/usage/training#custom-tokenizer" ```customize_tokenizer``` does not return anything. I tried to add ```return nlp``` or ```nlp.tokenizer``` nothing has changed.

```
python -m spacy debug data config_trf_ner.cfg --paths.train train.spacy --paths.dev dev.spacy  --code dil/data/training/functions.py

```

**Reply by mehmetilker** [reply]
I forget to mention. as an additional note, I am adding custom function the the config as follow:

```
[initialize.before_init]
@callbacks = "customize_tokenizer"
```
